File: coding-interview-sorting-algorithms-pro/src/main/java/com/example/sorting/problems/Problem4_SortColors.java
```java
package com.example.sorting.problems;

import com.example.sorting.utils.ArrayUtils;

import java.util.Arrays;

/**
 * Problem 4: Sort Colors (LeetCode 75 Equivalent)
 *
 * Given an array `nums` with `n` objects colored red, white, or blue, sort them in-place
 * so that objects of the same color are adjacent, with the colors in the order red, white, and blue.
 * Use the integers 0, 1, and 2 to represent red, white, and blue, respectively.
 *
 * You must solve this problem without using the library's sort function.
 *
 * Example:
 * Input: [2,0,2,1,1,0]
 * Output: [0,0,1,1,2,2]
 */
public class Problem4_SortColors {

    /**
     * Optimal Solution: Dutch National Flag Algorithm (Three-Way Partitioning)
     *
     * Algorithm:
     * This algorithm uses three pointers: `low`, `mid`, and `high`.
     * - `low`: Points to the end of the 0s section. All elements to the left of `low` (exclusive) are 0s.
     * - `mid`: Points to the current element being examined.
     * - `high`: Points to the beginning of the 2s section. All elements to the right of `high` (exclusive) are 2s.
     *
     * The invariant is:
     * - `nums[0...low-1]` contains all 0s.
     * - `nums[low...mid-1]` contains all 1s.
     * - `nums[mid...high]` contains unknown elements.
     * - `nums[high+1...n-1]` contains all 2s.
     *
     * Steps:
     * 1. Initialize `low = 0`, `mid = 0`, `high = nums.length - 1`.
     * 2. Iterate while `mid <= high`:
     *    a. If `nums[mid]` is 0:
     *       Swap `nums[low]` and `nums[mid]`.
     *       Increment both `low` and `mid`.
     *    b. If `nums[mid]` is 1:
     *       Increment `mid`.
     *    c. If `nums[mid]` is 2:
     *       Swap `nums[mid]` and `nums[high]`.
     *       Decrement `high`. (Note: `mid` is not incremented here because the element swapped from `high`
     *       could be 0, 1, or 2, and needs to be re-examined).
     *
     * Time Complexity: O(N) - Each element is visited and swapped at most a constant number of times.
     * Space Complexity: O(1) - In-place sorting.
     */
    public void sortColorsDutchNationalFlag(int[] nums) {
        if (nums == null || nums.length <= 1) {
            return;
        }

        int low = 0;    // Pointer for 0s (red)
        int mid = 0;    // Pointer for current element
        int high = nums.length - 1; // Pointer for 2s (blue)

        while (mid <= high) {
            switch (nums[mid]) {
                case 0: // If current element is 0 (red)
                    ArrayUtils.swap(nums, low, mid);
                    low++;
                    mid++;
                    break;
                case 1: // If current element is 1 (white)
                    mid++;
                    break;
                case 2: // If current element is 2 (blue)
                    ArrayUtils.swap(nums, mid, high);
                    high--;
                    // Note: mid is NOT incremented here because the element swapped from 'high'
                    // might be a 0 or 1, and needs to be re-examined.
                    break;
            }
        }
    }

    /**
     * Alternative Solution: Counting Sort (Two-Pass Approach)
     *
     * Algorithm:
     * Counting sort is efficient for sorting data within a small integer range.
     * 1. Count the occurrences of each color (0, 1, and 2) in the array.
     * 2. Overwrite the array based on the counts:
     *    - Fill the first `count[0]` positions with 0.
     *    - Fill the next `count[1]` positions with 1.
     *    - Fill the remaining `count[2]` positions with 2.
     *
     * Time Complexity: O(N) - One pass to count, one pass to fill the array.
     * Space Complexity: O(1) - As the range of numbers (0, 1, 2) is fixed and small, the count array size is constant (3).
     * Stability: Stable if implemented carefully, but for this problem, stability is not required.
     */
    public void sortColorsCountingSort(int[] nums) {
        if (nums == null || nums.length <= 1) {
            return;
        }

        // Assuming colors are only 0, 1, 2
        int[] counts = new int[3]; // counts[0] for 0s, counts[1] for 1s, counts[2] for 2s

        // First pass: Count occurrences of each color
        for (int num : nums) {
            if (num >= 0 && num <= 2) { // Basic validation
                counts[num]++;
            } else {
                throw new IllegalArgumentException("Array contains elements other than 0, 1, or 2.");
            }
        }

        // Second pass: Overwrite the array based on counts
        int index = 0;
        for (int i = 0; i < counts[0]; i++) {
            nums[index++] = 0;
        }
        for (int i = 0; i < counts[1]; i++) {
            nums[index++] = 1;
        }
        for (int i = 0; i < counts[2]; i++) {
            nums[index++] = 2;
        }
    }

    /**
     * Brute-Force Solution: Using Built-in Sort
     *
     * Algorithm:
     * Simply use the standard library's sorting function to sort the array.
     *
     * Time Complexity: O(N log N) - Standard library sorts are typically O(N log N).
     * Space Complexity: O(log N) to O(N) depending on the specific implementation (e.g., Java's `Arrays.sort()` for primitives is Dual-Pivot QuickSort with O(log N) space).
     * This solution violates the "without using the library's sort function" constraint but is a valid brute-force comparison.
     */
    public void sortColorsBuiltInSort(int[] nums) {
        if (nums == null || nums.length <= 1) {
            return;
        }
        Arrays.sort(nums);
    }
}
```

File: data-structures-sorting-algorithms-enhanced/sorting_algorithms_project/problems/problem_4_sort_colors.py
def sort_colors_two_pass_counting(nums):
    """
    Problem 4: Sort Colors (Dutch National Flag Problem) - Two-Pass Counting Sort.
    Given an array with n objects colored red, white, or blue (0, 1, or 2),
    sort them in-place.

    This approach uses counting sort logic. It counts the occurrences of each color
    and then overwrites the array based on these counts. This is not strictly
    in-place as it requires a pass to count and a pass to rewrite.

    Args:
        nums (list): The list of integers (0, 1, or 2) to be sorted.

    Time Complexity:
        - O(n) - Two passes over the array: one for counting, one for overwriting.

    Space Complexity:
        - O(1) - Uses a fixed amount of extra space (3 counters), independent of n.
    """
    if not nums:
        return

    counts = [0, 0, 0] # counts[0] for 0s, counts[1] for 1s, counts[2] for 2s

    # First pass: Count occurrences of each color
    for num in nums:
        counts[num] += 1

    # Second pass: Overwrite the original array based on counts
    index = 0
    # Place all 0s
    for _ in range(counts[0]):
        nums[index] = 0
        index += 1
    # Place all 1s
    for _ in range(counts[1]):
        nums[index] = 1
        index += 1
    # Place all 2s
    for _ in range(counts[2]):
        nums[index] = 2
        index += 1

    # The function modifies the array in-place.

def sort_colors_one_pass_three_pointers(nums):
    """
    Problem 4: Sort Colors (Dutch National Flag Problem) - One-Pass Three-Pointers (Optimal).
    This is the classic Dutch National Flag algorithm. It uses three pointers:
    - `low`: Points to the boundary between 0s and 1s. Elements before `low` are all 0s.
    - `mid`: The current element being considered.
    - `high`: Points to the boundary between 1s and 2s. Elements after `high` are all 2s.

    The algorithm iterates while `mid` is less than or equal to `high`.
    - If `nums[mid]` is 0, swap it with `nums[low]`, then increment `low` and `mid`.
    - If `nums[mid]` is 1, just increment `mid`.
    - If `nums[mid]` is 2, swap it with `nums[high]`, then decrement `high`.
      Note: `mid` is NOT incremented here because the swapped element from `high`
      needs to be re-evaluated (it could be 0, 1, or 2).

    Args:
        nums (list): The list of integers (0, 1, or 2) to be sorted.

    Time Complexity:
        - O(n) - Single pass over the array. Each element is visited at most twice (once by `mid`,
                 and potentially once if swapped with `low` or `high`).

    Space Complexity:
        - O(1) - All operations are done in-place without additional storage.
    """
    if not nums:
        return

    low = 0    # Pointer for the position where the next 0 should go
    mid = 0    # Current element being examined
    high = len(nums) - 1 # Pointer for the position where the next 2 should go

    while mid <= high:
        if nums[mid] == 0:
            # If current element is 0, swap it with element at `low` pointer
            # and move both `low` and `mid` pointers forward.
            nums[low], nums[mid] = nums[mid], nums[low]
            low += 1
            mid += 1
        elif nums[mid] == 1:
            # If current element is 1, it's already in its correct "middle" section,
            # so just move `mid` pointer forward.
            mid += 1
        else: # nums[mid] == 2
            # If current element is 2, swap it with element at `high` pointer
            # and move `high` pointer backward.
            # We do NOT increment `mid` here because the new element at `nums[mid]`
            # (which came from `high`) needs to be re-evaluated.
            nums[mid], nums[high] = nums[high], nums[mid]
            high -= 1

    # The function modifies the array in-place.
```